Windows build fails: posix_memalign and mmap not available on MinGW

## Problem

libvroom fails to compile on Windows with MinGW/gcc (Rtools), which is the standard R toolchain on Windows.

### Error 1: `posix_memalign` not available on MinGW

`include/libvroom/io_util.h:99` uses `posix_memalign` in the `#else` branch of `#ifdef _MSC_VER`. MinGW defines neither `_MSC_VER` nor provides `posix_memalign`, so compilation fails:

```
io_util.h:99:14: error: 'posix_memalign' was not declared in this scope
   99 |     int rc = posix_memalign(reinterpret_cast<void**>(&buf.data_), 64, buf.capacity_);
```

The same pattern affects the `_aligned_free` / `std::free` branch at line 120.

### Error 2: `mmap` / `munmap` not available on MinGW

`src/reader/mmap_source.cpp` uses `mmap()` and `munmap()` unconditionally with no Windows fallback. These POSIX functions don't exist on MinGW.

### Root Cause

The `#ifdef _MSC_VER` guards only handle MSVC but not MinGW, which is the gcc-based Windows compiler used by:
- R (via Rtools)
- Many other Windows C++ projects

### Suggested Fix

Use `#ifdef _WIN32` instead of `#ifdef _MSC_VER` for platform detection. `_WIN32` is defined by both MSVC and MinGW.

**io_util.h:**
```cpp
#ifdef _WIN32
    buf.data_ = static_cast<uint8_t*>(_aligned_malloc(buf.capacity_, 64));
#else
    int rc = posix_memalign(reinterpret_cast<void**>(&buf.data_), 64, buf.capacity_);
    ...
#endif
```

Same for the `_aligned_free` block.

**mmap_source.cpp:** Needs a Windows implementation using `CreateFileMapping` / `MapViewOfFile`, or a fallback to `read()` into an `AlignedBuffer`.

### Context

Discovered via R package CI: https://github.com/jimhester/vroom/actions/runs/21691610182/job/62552442646?pr=9

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -618,6 +618,13 @@ if(BUILD_TESTING)
     add_dependencies(io_util_test copy_test_data)
     gtest_discover_tests(io_util_test WORKING_DIRECTORY ${CMAKE_BINARY_DIR})
 
+    # MmapSource tests (memory-mapped file I/O)
+    add_executable(mmap_source_test test/mmap_source_test.cpp)
+    target_link_libraries(mmap_source_test PRIVATE vroom GTest::gtest_main pthread)
+    target_include_directories(mmap_source_test PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/include)
+    add_dependencies(mmap_source_test copy_test_data)
+    gtest_discover_tests(mmap_source_test WORKING_DIRECTORY ${CMAKE_BINARY_DIR})
+
     # CLI integration tests (tests the vroom command-line tool)
     add_executable(cli_test test/cli_test.cpp)
     target_link_libraries(cli_test PRIVATE GTest::gtest_main)

--- a/benchmark/basic_benchmarks.cpp
+++ b/benchmark/basic_benchmarks.cpp
@@ -101,14 +101,14 @@ static void BM_MemoryAllocation(benchmark::State& state) {
   size_t file_size = static_cast<size_t>(state.range(0));
 
   for (auto _ : state) {
-    void* ptr;
-    if (posix_memalign(&ptr, 64, file_size + LIBVROOM_PADDING) != 0) {
+    auto* data =
+        static_cast<uint8_t*>(libvroom::aligned_alloc_portable(file_size + LIBVROOM_PADDING));
+    if (!data) {
       state.SkipWithError("Failed to allocate memory");
       return;
     }
-    auto* data = static_cast<uint8_t*>(ptr);
     benchmark::DoNotOptimize(data);
-    std::free(data);
+    libvroom::aligned_free_portable(data);
   }
 
   state.SetBytesProcessed(static_cast<int64_t>(file_size * state.iterations()));
@@ -139,12 +139,11 @@ static void BM_WriteSequential(benchmark::State& state) {
   const size_t total_bytes = total_elements * sizeof(uint64_t);
 
   // Allocate aligned memory
-  void* ptr;
-  if (posix_memalign(&ptr, 64, total_bytes) != 0) {
+  auto* data = static_cast<uint64_t*>(libvroom::aligned_alloc_portable(total_bytes));
+  if (!data) {
     state.SkipWithError("Failed to allocate memory");
     return;
   }
-  auto* data = static_cast<uint64_t*>(ptr);
 
   for (auto _ : state) {
     // Sequential write: iterate through memory linearly
@@ -155,7 +154,7 @@ static void BM_WriteSequential(benchmark::State& state) {
     benchmark::ClobberMemory();
   }
 
-  std::free(data);
+  libvroom::aligned_free_portable(data);
 
   state.SetBytesProcessed(static_cast<int64_t>(total_bytes * state.iterations()));
   state.counters["Rows"] = static_cast<double>(rows);
@@ -177,12 +176,11 @@ static void BM_WriteStrided(benchmark::State& state) {
   const size_t stride = rows; // In elements (stride in bytes = rows * 8)
 
   // Allocate aligned memory
-  void* ptr;
-  if (posix_memalign(&ptr, 64, total_bytes) != 0) {
+  auto* data = static_cast<uint64_t*>(libvroom::aligned_alloc_portable(total_bytes));
+  if (!data) {
     state.SkipWithError("Failed to allocate memory");
     return;
   }
-  auto* data = static_cast<uint64_t*>(ptr);
 
   for (auto _ : state) {
     // Strided write: for each row, write fields with stride between columns
@@ -197,7 +195,7 @@ static void BM_WriteStrided(benchmark::State& state) {
     benchmark::ClobberMemory();
   }
 
-  std::free(data);
+  libvroom::aligned_free_portable(data);
 
   state.SetBytesProcessed(static_cast<int64_t>(total_bytes * state.iterations()));
   state.counters["Rows"] = static_cast<double>(rows);

--- a/benchmark/comparison_benchmarks.cpp
+++ b/benchmark/comparison_benchmarks.cpp
@@ -190,13 +190,11 @@ BENCHMARK(BM_parsing_approaches_quoted)
 // Memory bandwidth benchmark
 static void BM_memory_bandwidth(benchmark::State& state) {
   size_t size = static_cast<size_t>(state.range(0));
-  void* raw = nullptr;
-  int rc = posix_memalign(&raw, 64, size);
-  if (rc != 0 || !raw) {
+  auto* data = static_cast<char*>(libvroom::aligned_alloc_portable(size));
+  if (!data) {
     state.SkipWithError("Failed to allocate aligned memory");
     return;
   }
-  auto data = static_cast<char*>(raw);
 
   // Initialize data
   for (size_t i = 0; i < size; ++i) {
@@ -215,7 +213,7 @@ static void BM_memory_bandwidth(benchmark::State& state) {
 
   // Google Benchmark calculates throughput automatically
 
-  std::free(data);
+  libvroom::aligned_free_portable(data);
 }
 BENCHMARK(BM_memory_bandwidth)
     ->Range(1024, 1024 * 1024 * 100) // 1KB to 100MB

--- a/benchmark/transpose_benchmarks.cpp
+++ b/benchmark/transpose_benchmarks.cpp
@@ -361,15 +361,11 @@ static void transpose_nontemporal_store(const uint64_t* row_major, uint64_t* col
 #endif // __x86_64__
 
 // =============================================================================
-// Helper: posix_memalign-based allocation
+// Helper: portable aligned allocation for uint64_t arrays
 // =============================================================================
 
 static uint64_t* alloc_aligned_u64(size_t count) {
-  void* raw = nullptr;
-  int rc = posix_memalign(&raw, 64, count * sizeof(uint64_t));
-  if (rc != 0)
-    return nullptr;
-  return static_cast<uint64_t*>(raw);
+  return static_cast<uint64_t*>(libvroom::aligned_alloc_portable(count * sizeof(uint64_t)));
 }
 
 // =============================================================================
@@ -392,8 +388,8 @@ static void BM_TransposeSingleThreaded(benchmark::State& state) {
 
   if (!row_major || !col_major) {
     state.SkipWithError("Failed to allocate memory");
-    std::free(row_major);
-    std::free(col_major);
+    libvroom::aligned_free_portable(row_major);
+    libvroom::aligned_free_portable(col_major);
     return;
   }
 
@@ -419,8 +415,8 @@ static void BM_TransposeSingleThreaded(benchmark::State& state) {
   state.counters["MemoryMB"] =
       static_cast<double>(total_elements * sizeof(uint64_t) * 2) / (1024.0 * 1024.0);
 
-  std::free(row_major);
-  std::free(col_major);
+  libvroom::aligned_free_portable(row_major);
+  libvroom::aligned_free_portable(col_major);
 }
 
 /**
@@ -439,8 +435,8 @@ static void BM_TransposeMultiThreaded(benchmark::State& state) {
 
   if (!row_major || !col_major) {
     state.SkipWithError("Failed to allocate memory");
-    std::free(row_major);
-    std::free(col_major);
+    libvroom::aligned_free_portable(row_major);
+    libvroom::aligned_free_portable(col_major);
     return;
   }
 
@@ -463,8 +459,8 @@ static void BM_TransposeMultiThreaded(benchmark::State& state) {
   state.counters["Elements/s"] = benchmark::Counter(static_cast<double>(total_elements),
                                                     benchmark::Counter::kIsIterationInvariantRate);
 
-  std::free(row_major);
-  std::free(col_major);
+  libvroom::aligned_free_portable(row_major);
+  libvroom::aligned_free_portable(col_major);
 }
 
 /**
@@ -482,8 +478,8 @@ static void BM_TransposeBlocked(benchmark::State& state) {
 
   if (!row_major || !col_major) {
     state.SkipWithError("Failed to allocate memory");
-    std::free(row_major);
-    std::free(col_major);
+    libvroom::aligned_free_portable(row_major);
+    libvroom::aligned_free_portable(col_major);
     return;
   }
 
@@ -505,8 +501,8 @@ static void BM_TransposeBlocked(benchmark::State& state) {
   state.counters["Elements/s"] = benchmark::Counter(static_cast<double>(total_elements),
                                                     benchmark::Counter::kIsIterationInvariantRate);
 
-  std::free(row_major);
-  std::free(col_major);
+  libvroom::aligned_free_portable(row_major);
+  libvroom::aligned_free_portable(col_major);
 }
 
 /**
@@ -525,8 +521,8 @@ static void BM_TransposeBlockedMultiThreaded(benchmark::State& state) {
 
   if (!row_major || !col_major) {
     state.SkipWithError("Failed to allocate memory");
-    std::free(row_major);
-    std::free(col_major);
+    libvroom::aligned_free_portable(row_major);
+    libvroom::aligned_free_portable(col_major);
     return;
   }
 
@@ -549,8 +545,8 @@ static void BM_TransposeBlockedMultiThreaded(benchmark::State& state) {
   state.counters["Elements/s"] = benchmark::Counter(static_cast<double>(total_elements),
                                                     benchmark::Counter::kIsIterationInvariantRate);
 
-  std::free(row_major);
-  std::free(col_major);
+  libvroom::aligned_free_portable(row_major);
+  libvroom::aligned_free_portable(col_major);
 }
 
 /**
@@ -572,8 +568,8 @@ static void BM_TransposeScaling(benchmark::State& state) {
 
   if (!row_major || !col_major) {
     state.SkipWithError("Failed to allocate memory");
-    std::free(row_major);
-    std::free(col_major);
+    libvroom::aligned_free_portable(row_major);
+    libvroom::aligned_free_portable(col_major);
     return;
   }
 
@@ -617,8 +613,8 @@ static void BM_TransposeScaling(benchmark::State& state) {
   state.counters["Elements/s"] = benchmark::Counter(static_cast<double>(total_elements),
                                                     benchmark::Counter::kIsIterationInvariantRate);
 
-  std::free(row_major);
-  std::free(col_major);
+  libvroom::aligned_free_portable(row_major);
+  libvroom::aligned_free_portable(col_major);
 }
 
 // =============================================================================
@@ -899,8 +895,8 @@ static void BM_TransposeSIMD(benchmark::State& state) {
 
   if (!row_major || !col_major) {
     state.SkipWithError("Failed to allocate memory");
-    std::free(row_major);
-    std::free(col_major);
+    libvroom::aligned_free_portable(row_major);
+    libvroom::aligned_free_portable(col_major);
     return;
   }
 
@@ -959,8 +955,8 @@ static void BM_TransposeSIMD(benchmark::State& state) {
   state.counters["Elements/s"] = benchmark::Counter(static_cast<double>(total_elements),
                                                     benchmark::Counter::kIsIterationInvariantRate);
 
-  std::free(row_major);
-  std::free(col_major);
+  libvroom::aligned_free_portable(row_major);
+  libvroom::aligned_free_portable(col_major);
 }
 
 // SIMD method comparison at key sizes
@@ -1018,8 +1014,8 @@ static void BM_TransposeSIMD_MT(benchmark::State& state) {
 
   if (!row_major || !col_major) {
     state.SkipWithError("Failed to allocate memory");
-    std::free(row_major);
-    std::free(col_major);
+    libvroom::aligned_free_portable(row_major);
+    libvroom::aligned_free_portable(col_major);
     return;
   }
 
@@ -1065,8 +1061,8 @@ static void BM_TransposeSIMD_MT(benchmark::State& state) {
   state.counters["Elements/s"] = benchmark::Counter(static_cast<double>(total_elements),
                                                     benchmark::Counter::kIsIterationInvariantRate);
 
-  std::free(row_major);
-  std::free(col_major);
+  libvroom::aligned_free_portable(row_major);
+  libvroom::aligned_free_portable(col_major);
 }
 
 // Multi-threaded comparison

--- a/include/libvroom/io_util.h
+++ b/include/libvroom/io_util.h
@@ -25,6 +25,10 @@
 #include <stdexcept>
 #include <string>
 
+#ifdef _WIN32
+#include <malloc.h> // _aligned_malloc, _aligned_free
+#endif
+
 namespace libvroom {
 
 /**
@@ -93,7 +97,7 @@ public:
     buf.capacity_ = size + padding;
     buf.size_ = size;
 
-#ifdef _MSC_VER
+#ifdef _WIN32
     buf.data_ = static_cast<uint8_t*>(_aligned_malloc(buf.capacity_, 64));
 #else
     int rc = posix_memalign(reinterpret_cast<void**>(&buf.data_), 64, buf.capacity_);
@@ -117,7 +121,7 @@ public:
 private:
   void free_data() {
     if (data_) {
-#ifdef _MSC_VER
+#ifdef _WIN32
       _aligned_free(data_);
 #else
       std::free(data_);
@@ -132,6 +136,38 @@ private:
   size_t size_ = 0;
   size_t capacity_ = 0;
 };
+
+/**
+ * @brief Allocate raw aligned memory (64-byte alignment).
+ *
+ * Portable wrapper around platform-specific aligned allocation.
+ * Caller must free with aligned_free_portable().
+ *
+ * @param size Number of bytes to allocate.
+ * @return Pointer to allocated memory, or nullptr on failure.
+ */
+inline void* aligned_alloc_portable(size_t size) {
+#ifdef _WIN32
+  return _aligned_malloc(size, 64);
+#else
+  void* ptr = nullptr;
+  if (posix_memalign(&ptr, 64, size) != 0) {
+    return nullptr;
+  }
+  return ptr;
+#endif
+}
+
+/**
+ * @brief Free memory allocated with aligned_alloc_portable().
+ */
+inline void aligned_free_portable(void* ptr) {
+#ifdef _WIN32
+  _aligned_free(ptr);
+#else
+  std::free(ptr);
+#endif
+}
 
 /**
  * @brief Load a file into an aligned buffer.

--- a/src/reader/mmap_source.cpp
+++ b/src/reader/mmap_source.cpp
@@ -1,5 +1,113 @@
 #include "libvroom/vroom.h"
 
+#ifdef _WIN32
+
+// =============================================================================
+// Windows implementation using CreateFileMapping / MapViewOfFile
+// =============================================================================
+
+#ifndef NOMINMAX
+#define NOMINMAX
+#endif
+#include <windows.h>
+
+namespace libvroom {
+
+struct MmapSource::Impl {
+  HANDLE file_handle = INVALID_HANDLE_VALUE;
+  HANDLE mapping_handle = nullptr;
+  const char* data = nullptr;
+  size_t size = 0;
+};
+
+MmapSource::MmapSource() : impl_(std::make_unique<Impl>()) {}
+
+MmapSource::~MmapSource() {
+  close();
+}
+
+Result<bool> MmapSource::open(const std::string& path) {
+  if (impl_->file_handle != INVALID_HANDLE_VALUE) {
+    close();
+  }
+
+  impl_->file_handle = CreateFileA(path.c_str(), GENERIC_READ, FILE_SHARE_READ, nullptr,
+                                   OPEN_EXISTING, FILE_FLAG_SEQUENTIAL_SCAN, nullptr);
+  if (impl_->file_handle == INVALID_HANDLE_VALUE) {
+    return Result<bool>::failure("Failed to open file: " + path);
+  }
+
+  LARGE_INTEGER file_size;
+  if (!GetFileSizeEx(impl_->file_handle, &file_size)) {
+    CloseHandle(impl_->file_handle);
+    impl_->file_handle = INVALID_HANDLE_VALUE;
+    return Result<bool>::failure("Failed to get file size: " + path);
+  }
+
+  impl_->size = static_cast<size_t>(file_size.QuadPart);
+
+  if (impl_->size == 0) {
+    // Empty file - no need to map
+    return Result<bool>::success(true);
+  }
+
+  impl_->mapping_handle =
+      CreateFileMappingA(impl_->file_handle, nullptr, PAGE_READONLY, 0, 0, nullptr);
+  if (impl_->mapping_handle == nullptr) {
+    CloseHandle(impl_->file_handle);
+    impl_->file_handle = INVALID_HANDLE_VALUE;
+    return Result<bool>::failure("Failed to create file mapping: " + path);
+  }
+
+  void* ptr = MapViewOfFile(impl_->mapping_handle, FILE_MAP_READ, 0, 0, 0);
+  if (ptr == nullptr) {
+    CloseHandle(impl_->mapping_handle);
+    impl_->mapping_handle = nullptr;
+    CloseHandle(impl_->file_handle);
+    impl_->file_handle = INVALID_HANDLE_VALUE;
+    return Result<bool>::failure("Failed to map view of file: " + path);
+  }
+
+  impl_->data = static_cast<const char*>(ptr);
+  return Result<bool>::success(true);
+}
+
+const char* MmapSource::data() const {
+  return impl_->data;
+}
+
+size_t MmapSource::size() const {
+  return impl_->size;
+}
+
+bool MmapSource::is_open() const {
+  return impl_->file_handle != INVALID_HANDLE_VALUE;
+}
+
+void MmapSource::close() {
+  if (impl_->data != nullptr) {
+    UnmapViewOfFile(impl_->data);
+    impl_->data = nullptr;
+  }
+  if (impl_->mapping_handle != nullptr) {
+    CloseHandle(impl_->mapping_handle);
+    impl_->mapping_handle = nullptr;
+  }
+  if (impl_->file_handle != INVALID_HANDLE_VALUE) {
+    CloseHandle(impl_->file_handle);
+    impl_->file_handle = INVALID_HANDLE_VALUE;
+  }
+  impl_->size = 0;
+}
+
+} // namespace libvroom
+
+#else // POSIX
+
+// =============================================================================
+// POSIX implementation using mmap / munmap
+// =============================================================================
+
 #include <fcntl.h>
 #include <sys/mman.h>
 #include <sys/stat.h>
@@ -83,3 +191,5 @@ void MmapSource::close() {
 }
 
 } // namespace libvroom
+
+#endif // _WIN32

--- a/test/mmap_source_test.cpp
+++ b/test/mmap_source_test.cpp
@@ -1,0 +1,139 @@
+/**
+ * @file mmap_source_test.cpp
+ * @brief Tests for MmapSource memory-mapped file I/O.
+ *
+ * Tests open, close, content integrity, empty files, error handling,
+ * and reopen behavior for the platform-specific MmapSource implementation.
+ *
+ * @see GitHub issue #649
+ */
+
+#include "libvroom/vroom.h"
+
+#include <cstring>
+#include <fstream>
+#include <gtest/gtest.h>
+#include <string>
+
+using libvroom::MmapSource;
+
+class MmapSourceTest : public ::testing::Test {
+protected:
+  std::string testDataPath(const std::string& subpath) { return "test/data/" + subpath; }
+};
+
+// =============================================================================
+// Open and basic state tests
+// =============================================================================
+
+TEST_F(MmapSourceTest, DefaultStateNotOpen) {
+  MmapSource source;
+  EXPECT_FALSE(source.is_open());
+  EXPECT_EQ(source.size(), 0);
+  EXPECT_EQ(source.data(), nullptr);
+}
+
+TEST_F(MmapSourceTest, OpenValidFile) {
+  std::string path = testDataPath("basic/simple.csv");
+  std::ifstream check(path);
+  if (!check.good()) {
+    GTEST_SKIP() << "Test data not found: " << path;
+  }
+
+  MmapSource source;
+  auto result = source.open(path);
+  ASSERT_TRUE(result.ok);
+  EXPECT_TRUE(source.is_open());
+  EXPECT_GT(source.size(), 0);
+  EXPECT_NE(source.data(), nullptr);
+}
+
+TEST_F(MmapSourceTest, OpenNonExistentFile) {
+  MmapSource source;
+  auto result = source.open("nonexistent_file_that_does_not_exist.csv");
+  EXPECT_FALSE(result.ok);
+  EXPECT_FALSE(source.is_open());
+}
+
+// =============================================================================
+// Content integrity tests
+// =============================================================================
+
+TEST_F(MmapSourceTest, ContentMatchesFile) {
+  std::string path = testDataPath("basic/simple.csv");
+  std::ifstream file(path, std::ios::binary);
+  if (!file.good()) {
+    GTEST_SKIP() << "Test data not found: " << path;
+  }
+  std::string expected((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+
+  MmapSource source;
+  ASSERT_TRUE(source.open(path).ok);
+  EXPECT_EQ(source.size(), expected.size());
+  EXPECT_EQ(std::memcmp(source.data(), expected.data(), expected.size()), 0);
+}
+
+TEST_F(MmapSourceTest, EmptyFile) {
+  std::string temp = "/tmp/libvroom_mmap_test_empty.csv";
+  { std::ofstream ofs(temp, std::ios::binary); }
+
+  MmapSource source;
+  auto result = source.open(temp);
+  ASSERT_TRUE(result.ok);
+  EXPECT_TRUE(source.is_open());
+  EXPECT_EQ(source.size(), 0);
+
+  std::remove(temp.c_str());
+}
+
+// =============================================================================
+// Close and lifecycle tests
+// =============================================================================
+
+TEST_F(MmapSourceTest, CloseReleasesResources) {
+  MmapSource source;
+  ASSERT_TRUE(source.open(testDataPath("basic/simple.csv")).ok);
+  EXPECT_TRUE(source.is_open());
+
+  source.close();
+  EXPECT_FALSE(source.is_open());
+  EXPECT_EQ(source.size(), 0);
+}
+
+TEST_F(MmapSourceTest, DoubleCloseIsSafe) {
+  MmapSource source;
+  ASSERT_TRUE(source.open(testDataPath("basic/simple.csv")).ok);
+  source.close();
+  source.close(); // Should not crash
+  EXPECT_FALSE(source.is_open());
+}
+
+TEST_F(MmapSourceTest, ReopenDifferentFile) {
+  std::string path1 = testDataPath("basic/simple.csv");
+  std::string path2 = testDataPath("quoted/quoted_fields.csv");
+
+  std::ifstream check1(path1), check2(path2);
+  if (!check1.good() || !check2.good()) {
+    GTEST_SKIP() << "Test data files not found";
+  }
+
+  MmapSource source;
+  ASSERT_TRUE(source.open(path1).ok);
+  size_t first_size = source.size();
+  EXPECT_GT(first_size, 0);
+
+  // Opening a new file should implicitly close the first
+  ASSERT_TRUE(source.open(path2).ok);
+  EXPECT_TRUE(source.is_open());
+  EXPECT_GT(source.size(), 0);
+}
+
+TEST_F(MmapSourceTest, DestructorCleansUp) {
+  // Construct and open in a scope, let destructor run
+  {
+    MmapSource source;
+    ASSERT_TRUE(source.open(testDataPath("basic/simple.csv")).ok);
+    EXPECT_TRUE(source.is_open());
+  }
+  // No crash or leak — destructor handled cleanup
+}

